apache-kyuubi 1.6.1 (new formula)

yaooqinn opened the pull request:
<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

chenrui333 commented on Formula/kyuubi.rb:
```diff
@@ -0,0 +1,36 @@
+class Kyuubi < Formula
+  desc "Serverless SQL on data warehouses and lakehouses"
+  homepage "https://kyuubi.apache.org/"
+  version "1.6.1-incubating"
+  url "https://dlcdn.apache.org/kyuubi/kyuubi-#{version}/apache-kyuubi-#{version}-bin.tgz"
```
let's maybe do in this way, consistent with the other apache formulae
```suggestion
  url "https://www.apache.org/dyn/closer.lua?path=kyuubi/1.6.1/apache-kyuubi-1.6.1-incubating-bin.tar.gz"
  mirror "https://archive.apache.org/dist/kyuubi/kyuubi-1.6.1-incubating/apache-kyuubi-1.6.1-incubating-bin.tgz"
  version "1.6.1"
```

chenrui333 commented on Formula/kyuubi.rb:
```diff
@@ -0,0 +1,36 @@
+  license "Apache-2.0"
+  head "https://github.com/apache/kyuubi.git", branch: "master"
+
+  depends_on "openjdk@8"
```
Does this work with the other jdk? If yes, we might want to do something like this

```
JAVA_HOME:     Language::Java.overridable_java_home_env("1.8")[:JAVA_HOME],
```

chenrui333 commented on Formula/kyuubi.rb:
```diff
@@ -0,0 +1,36 @@
+
+  def install
+    libexec.install Dir["*"]
+    (var/"kyuubi/logs").mkpath
+    (var/"kyuubi/work").mkpath
+    (var/"kyuubi/pid").mkpath
```
this can goes to `post_install`

chenrui333 commented on Formula/kyuubi.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  test do
+    assert_match "org.apache.kyuubi.server.KyuubiServer", shell_output("#{bin}/kyuubi stop")
```
there are other examples in the repo that can help you with the server start/stop testing.

yaooqinn replied:
Addressed other comments, will reference some examples to improve the testing tomorrow(UTC+8)

SMillerDev commented on Formula/kyuubi.rb:
```diff
@@ -0,0 +1,52 @@
+  license "Apache-2.0"
+  head "https://github.com/apache/kyuubi.git", branch: "master"
+
+  def create_bin_with_envs(target = "kyuubi", source = "kyuubi")
+    (bin/target).write_env_script libexec/"bin"/source,
+      JAVA_HOME:            Language::Java.overridable_java_home_env("1.8")[:JAVA_HOME],
+      SPARK_HOME:           "${SPARK_HOME:-#{Formula["apache-spark"].opt_libexec}}",
+      KYUUBI_HOME:          libexec.to_s,
+      KYUUBI_LOG_DIR:       "${KYUUBI_LOG_DIR:-#{var}/kyuubi/logs}",
+      KYUUBI_PID_DIR:       "${KYUUBI_PID_DIR:-#{var}/kyuubi/pid}",
+      KYUUBI_WORK_DIR_ROOT: "${KYUUBI_WORK_DIR_ROOT:-#{var}/kyuubi/work}"
+  end
```
You can put this in the install block. Just make the ENV a variable

yaooqinn replied:
Can you give me an example?

When I did something like `ENV["JAVA_HOME"] = Formula["openjdk@8"].opt_prefix` in install block. the scripts of kyuubi in `bin` did not respect it

SMillerDev replied:
```rb
setting = {
      JAVA_HOME:            Language::Java.overridable_java_home_env("1.8")[:JAVA_HOME],
      SPARK_HOME:           "${SPARK_HOME:-#{Formula["apache-spark"].opt_libexec}}",
      KYUUBI_HOME:          libexec.to_s,
      KYUUBI_LOG_DIR:       "${KYUUBI_LOG_DIR:-#{var}/kyuubi/logs}",
      KYUUBI_PID_DIR:       "${KYUUBI_PID_DIR:-#{var}/kyuubi/pid}",
      KYUUBI_WORK_DIR_ROOT: "${KYUUBI_WORK_DIR_ROOT:-#{var}/kyuubi/work}"
}
(bin/"a").write_env_script libexec/"bin/b", setting
(bin/"c").write_env_script libexec/"bin/d", setting
```

yaooqinn replied:
thanks for the advice @SMillerDev

SMillerDev commented on Formula/kyuubi.rb:
```diff
@@ -0,0 +1,52 @@
+      KYUUBI_PID_DIR:       "${KYUUBI_PID_DIR:-#{var}/kyuubi/pid}",
+      KYUUBI_WORK_DIR_ROOT: "${KYUUBI_WORK_DIR_ROOT:-#{var}/kyuubi/work}"
+  end
+
```
If it depends on Java and apache-spark you should add those as dependencies

yaooqinn replied:
kyuubi and spark currently default to java 8 for compilation, spark uses java 19 in brew but kyuubi currently guarantees up to java 11 in runtime. I can not pass the `brew audit` verification.

yaooqinn replied:
Narrowed the scope of spark dep to test to fix this. Spark's one of those SQL engines that kyuubi supports. Users are free to use kyuubi with Apache Flink, Hive, Doris; Trino, etc.

SMillerDev commented on Formula/kyuubi.rb:
```diff
@@ -0,0 +1,53 @@
+  license "Apache-2.0"
+  head "https://github.com/apache/kyuubi.git", branch: "master"
+
+  depends_on "apache-spark" => :test
```
You're hardcoding the path in the install block. It's not test-only

yaooqinn replied:
Does it work if I move spark home to the test part? I verified locally and it said `Error: kyuubi is missing test dependencies: apache-spark`. How can I make it automatically install test dependencies?

yaooqinn replied:
seems OK now

chenrui333 replied:
yeah, that is way to install test dependency, you did correctly.

yaooqinn replied:
Thank you for the confirmation

SMillerDev commented on Formula/kyuubi.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  test do
+    ENV.append "_JAVA_OPTIONS", "-Duser.home=#{testpath}"
+    ENV.append "_JAVA_OPTIONS", "-Dkyuubi.frontend.bind.host=localhost"
+    ENV.append "_JAVA_OPTIONS", "-Dkyuubi.zookeeper.embedded.data.dir=#{testpath}/zk"
+    ENV.append "_JAVA_OPTIONS", "-Dkyuubi.metrics.reporters=''"
```
Will users need to set all these too?

yaooqinn replied:
not necessary

SMillerDev replied:
Then why are they required in this block?

yaooqinn replied:
kyuubi.frontend.bind.host is hardcoded to localhost for convenience.

kyuubi.zookeeper.embedded.data.dir is set to testpath to avoid polluting the one created by a user.

kyuubi.metrics.reporters is disabled for tests, which reports to a local json file by default,

SMillerDev replied:
Can we remove anything that isn't absolutely required to run the test. All the extra settings are making it difficult to see how the command would work outside the test block.

chenrui333 replied:
like @SMillerDev said, can we create a simple config in the kyuubi project and included in here for testing? That might be better and easy to maintain.

yaooqinn replied:
The custom test configs are a subset of the runtime one now. only `localhost` is used for convenience

yaooqinn commented:
Any more concerns？ @SMillerDev @chenrui333

chenrui333 commented:
@yaooqinn also maybe good to squash the commits into fewer commits so that they can also be read. 👍 

This PR is kind of down the PR chain, and feel free to create a new one to get better PR triage. (after the PR feedback got addressed) Thanks!!

yaooqinn commented:
Thank you, and sorry for the late reply. I was waiting for the apache kyuubi 1.7.0 release.

Addressed the comments to remove unnecessary settings for the test phase. Please take a look when you have time.

chenrui333 commented on Formula/kyuubi.rb:
```diff
@@ -0,0 +1,48 @@
+  end
+
+  test do
+    ENV.append "_JAVA_OPTIONS", "-Duser.home=#{testpath}"
+    ENV.append "_JAVA_OPTIONS", "-Dkyuubi.frontend.bind.host=localhost"
+    ENV.append "KYUUBI_WORK_DIR_ROOT", testpath/"work"
+    ENV.append "SPARK_HOME", Formula["apache-spark"].opt_libexec.to_s
```
looks like the config setup in here has not been addressed?

Additionally, if the formula needs apache-spark to run, maybe it should be hard dependency on it instead of just test dependency?

yaooqinn replied:
We also support apache/hive, apache/flink, apache/doris, presto, and among others. apache/spark is one option.

Besides，https://github.com/Homebrew/homebrew-core/pull/123106#discussion_r1105456926 shows that `brew audit` may be too strict for JDK dependency check.
